### PR TITLE
[SYCLomatic][CMake] Emit note msgs when detecting conflicts of CMake migration rules

### PR DESCRIPTION
**Pull request by tomflinda**
1. Fix the conflicts in current CMake migration rules set
2. Further refine the CMake migration rules
3. Add some print statements for debug purpose

### CODE REVIEW

File: clang/lib/DPCT/MigrateCmakeScript.cpp [hunk 1]
```diff
@@ -737,18 +737,35 @@ void doCmakeScriptMigration(const clang::tooling::UnifiedPath &InRoot,
...
+    DuplicateFilter.insert(R.CmakeSyntax);
+
+    auto PR =
+        MetaRuleObject::PatternRewriter(R.In, R.Out, R.Subrules, R.MatchMode,
+                                        R.RuleId, R.CmakeSyntax, R.Priority);
+
+    auto Iter = CmakeBuildInRules.find(PR.CmakeSyntax);
+    if (Iter != CmakeBuildInRules.end()) {
+      if (PR.Priority == RulePriority::Takeover) {
+        assert(Iter->second.Priority < PR.Priority &&
```
Comment (zhiweij1) [inline clang/lib/DPCT/MigrateCmakeScript.cpp]: How about printing the rule ID of the duplicated rules?

File: clang/lib/DPCT/MigrateCmakeScript.cpp [hunk 2]
```diff
@@ -737,18 +737,35 @@ void doCmakeScriptMigration(const clang::tooling::UnifiedPath &InRoot,
...
-  auto PR =
-      MetaRuleObject::PatternRewriter(R.In, R.Out, R.Subrules, R.MatchMode,
-                                      R.RuleId, R.CmakeSyntax, R.Priority);
-
-  auto Iter = CmakeBuildInRules.find(PR.CmakeSyntax);
-  if (Iter != CmakeBuildInRules.end()) {
-    if (PR.Priority == RulePriority::Takeover) {
-      assert(Iter->second.Priority < PR.Priority &&
-             "Two same cmake syntaxes have \'Takeover\' priority.\n ");
+
```
Comment (zhiweij1) [inline clang/lib/DPCT/MigrateCmakeScript.cpp]:
```suggestion
```

File: clang/lib/DPCT/MigrateCmakeScript.cpp (hunk 1 above)
Comment (tomflinda) [reply clang/lib/DPCT/MigrateCmakeScript.cpp]: okay

**Review by zhiweij1 — APPROVED** [review]

File: clang/lib/DPCT/MigrateCmakeScript.cpp [hunk 3]
```diff
@@ -737,18 +737,41 @@ void doCmakeScriptMigration(const clang::tooling::UnifiedPath &InRoot,
...
+#ifdef __DEBUG_CMAKE_SCRIPR_MIGRATION
```
Comment (zhimingwang36) [inline clang/lib/DPCT/MigrateCmakeScript.cpp]: Please remove the debug purpose code.

File: clang/lib/DPCT/MigrateCmakeScript.cpp [hunk 4]
```diff
@@ -737,18 +737,41 @@ void doCmakeScriptMigration(const clang::tooling::UnifiedPath &InRoot,
...
+      if (PR.Priority == RulePriority::Takeover) {
+        if (Iter->second.Priority < PR.Priority) {
+          CmakeBuildInRules[PR.CmakeSyntax] = PR;
+        } else {
+          llvm::outs() << " [NOTE] Two Yaml rules with same CMake syntax have "
+                          "\'Takeover\' priority:\n"
+                       << "\tNew rule id:\"" << R.RuleId << "\"\n"
+                       << "\tExiting rule id:\"" << Iter->second.RuleId
+                       << "\"\n"
+                       << " The new rule has been skipped.\n";
```
Comment (zhimingwang36) [inline clang/lib/DPCT/MigrateCmakeScript.cpp]: Let's keep the message simple in single line:  
[Warnning]: Two migration rules (Rule_id_1: source of the rule,  Rule_id:source of the rule ) are duplicated, the migrtion rule (Rule_id: source of the rule) is ingored.

File: clang/lib/DPCT/MigrateCmakeScript.cpp [hunk 5]
```diff
@@ -737,18 +737,41 @@ void doCmakeScriptMigration(const clang::tooling::UnifiedPath &InRoot,
...
+        }
+      } else {
```
Comment (zhimingwang36) [inline clang/lib/DPCT/MigrateCmakeScript.cpp]: where is the test case for this wanning ?

**Review by zhimingwang36 — CHANGES_REQUESTED** [review]

File: clang/lib/DPCT/MigrateCmakeScript.cpp (hunk 5 above)
Comment (tomflinda) [reply clang/lib/DPCT/MigrateCmakeScript.cpp]: I planned to add a behavior test in the SYCLomatic test repo; while a lit test is also workable, I will update it.

File: clang/lib/DPCT/MigrateCmakeScript.cpp (hunk 4 above)
Comment (tomflinda) [reply clang/lib/DPCT/MigrateCmakeScript.cpp]: Okay.

File: clang/lib/DPCT/MigrateCmakeScript.cpp (hunk 3 above)
Comment (tomflinda) [reply clang/lib/DPCT/MigrateCmakeScript.cpp]: okay

**Review by ShengchenJ — APPROVED** [review]

File: clang/lib/DPCT/MigrateCmakeScript.cpp [hunk 6]
```diff
@@ -743,10 +743,14 @@ void registerCmakeMigrationRule(MetaRuleObject &R) {
...
   auto Iter = CmakeBuildInRules.find(PR.CmakeSyntax);
   if (Iter != CmakeBuildInRules.end()) {
-    if (PR.Priority == RulePriority::Takeover) {
-      assert(Iter->second.Priority < PR.Priority &&
-             "Two same cmake syntaxes have \'Takeover\' priority.\n ");
+    if (PR.Priority == RulePriority::Takeover &&
+        Iter->second.Priority < PR.Priority) {
       CmakeBuildInRules[PR.CmakeSyntax] = PR;
+    } else {
+      llvm::outs() << "[Warnning]: Two migration rules (Rule_id_1:" << R.RuleId
```
Comment (zhimingwang36) [inline clang/lib/DPCT/MigrateCmakeScript.cpp]:
```suggestion
      llvm::outs() << "[Warnning]: Two migration rules (Rule:" << R.RuleId
```

File: clang/lib/DPCT/MigrateCmakeScript.cpp [hunk 7]
```diff
@@ -743,10 +743,14 @@ void registerCmakeMigrationRule(MetaRuleObject &R) {
...
+                   << ", Rule_id_2:" << Iter->second.RuleId
+                   << ") are duplicated, the migrtion rule (Rule_id:"
```
Comment (zhimingwang36) [inline clang/lib/DPCT/MigrateCmakeScript.cpp]:
```suggestion
                   << ") are duplicated, the migrtion rule (Rule:"
```

File: clang/lib/DPCT/MigrateCmakeScript.cpp [hunk 8]
```diff
@@ -743,10 +743,14 @@ void registerCmakeMigrationRule(MetaRuleObject &R) {
...
+                   << R.RuleId << ") is ingored.\n";
```
Comment (zhimingwang36) [inline clang/lib/DPCT/MigrateCmakeScript.cpp]:
```suggestion
                   << R.RuleId << ") is ignored.\n";
```

**Review by zhimingwang36 — APPROVED** [review]
LGTM

**Pull request merged**